File: (4)reConstructBinaryTree/Solution.java
 /*
 题目描述

输入某二叉树的前序遍历和中序遍历的结果，请重建出该二叉树。
假设输入的前序遍历和中序遍历的结果中都不含重复的数字。例
如输入前序遍历序列{1,2,4,7,3,5,6,8}和中序遍历序列{4,7,2,1,5,3,8,6}，
则重建二叉树并返回。
 */
 class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode(int x) { val = x; }
  }
 
public class Solution {

    public TreeNode reConstructBinaryTree(int [] pre,int [] in) {
    	int flag = 0;
        for(int i=0; i<in.length; i++){
        	if(in[i]==pre[0]){
        		flag = i;
        	}
        }
    	        TreeNode node = new TreeNode(in[flag]);
    	        node.left=null;
    	        node.right=null;
    	        if(flag>0){
    	        	int [] leftPre = getLeftPre(flag,pre);
        		int [] leftIn = getLeftIn(flag,in);
        		        		node.left=reConstructBinaryTree(leftPre,leftIn);

    	        }
    	        if(flag<in.length-1){
    	        	int [] rightPre = getRightPre(flag,pre);
        		int [] rightIn = getRightIn(flag,in);
        		
        		node.right=reConstructBinaryTree(rightPre,rightIn);
    	        }
        		return node;
    }

    public int [] getRightPre(int i, int [] pre){
    	int num;
    	if(pre.length!=i)
    		 num = pre.length-i-1;
    	else num = pre.length-i;
    		 int [] rightPre = new int [num];
    	for(int j=i+1; j<pre.length; j++){
    		rightPre[j-i-1] = pre[j];
    	}
     	return rightPre;
    }

    public int [] getLeftPre(int i, int [] pre){
    	int [] leftPre= new int [i];
    	for(int j=1; j<=i; j++){
    		leftPre[j-1]=pre[j];
    	}
    	return leftPre;
    }

    public int [] getRightIn(int i, int [] in){
    	int num;
    	if(in.length!=i)
    		 num = in.length-i-1;
    	else num = in.length-i;
    	int [] rightIn = new int[num];
    	for(int j=i+1; j<in.length; j++){
    		rightIn[j-i-1]=in[j];
    	}
    	return rightIn;
    }

    public int [] getLeftIn(int i, int [] in){
    	int [] leftIn = new int[i];
    	for(int j=0; j<i; j++){
    		leftIn[j]=in[j];
    	}
    	return leftIn;
    }

    public static void main(String []args){
    	Solution s = new Solution();
    	int [] pre={1,2,4,7,3,5,6,8};
    	int [] in={4,7,2,1,5,3,8,6};
    	System.out.println(s.reConstructBinaryTree(pre,in).val);
    }
}
/*
1.寻找根节点
2.根据根节点判断出左右子树，然后将左右子树当做一棵树进行递归求解
3.理清思路，列出边界情况
*/

File: (4)reConstructBinaryTree/Solution.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* reConstructBinaryTree(vector<int> pre,vector<int> vin) {
        if(pre.empty() || vin.empty()) {
            return NULL;
        }
        return constucetBinaryTree(0, pre.size()-1, 0, vin.size()-1, pre, vin);
    }

    public:
    TreeNode* constucetBinaryTree(int preHead, int preTail, int vinHead,int vinTail, vector<int> pre, vector<int> vin)  {
        if(preHead > preTail || vinHead > vinTail) 
            return NULL;

        TreeNode* treeNode = new TreeNode(pre[preHead]);
        int index = preHead;
        for (int i = vinHead; i <= vinTail; i++) {
            if(vin[i] == pre[preHead]) {
                index = i;
                break;
            }
        }
        
        treeNode->left = constucetBinaryTree(preHead+1, preHead + index - vinHead, vinHead, index-1, pre, vin);
        treeNode->right = constucetBinaryTree(preHead + index - vinHead + 1, preTail, index+1, vinTail, pre, vin);
        return treeNode;
    }
    

};


举一反三，通过这一个题目，我们便可以解决所有的二叉树重新构建问题，对于树的问题，由于该数据结构自身的特性，我们很容易将大问题
分解为子问题，通过递归的方式来解决问题，首先是获取到子节点，然后构建其左右子树，关键节点在于对其中递归点的切分，首先切分出来
从何处递归操作，这里要注意对于其中索引位置的控制，是很容易出现问题的地方。